rust/src/mainthread.rs

```diff
@@ -0,0 +1,204 @@
+//! This module provides two ways to execute "jobs":
+//!
+//! 1. On the Binary Ninja main thread (the UI event thread when running in the GUI application):
+//!    * [execute_on_main_thread]
+//!    * [execute_on_main_thread_and_wait]
+//! 2. On a worker thread
+//!
+//! Any manipulation of the GUI should be performed on the main thread, but any
+//! non-GUI work is generally better to be performed using a worker. This is
+//! especially true for any longer-running work, as the user interface will
+//! be unable to update itself while a job is executing on the main thread.
+//!
+//! There are three worker queues, in order of decreasing priority:
+//!
+//!    1. The Interactive Queue ([worker_interactive_enqueue])
+//!    2. The Priority Queue ([worker_priority_enqueue])
+//!    3. The Worker Queue ([worker_enqueue])
+//!
+//! All of these queues are serviced by the same pool of worker threads. The
+//! difference between the queues is basically one of priority: one queue must
+//! be empty of jobs before a worker thread will execute a job from a lower
+//! priority queue.
+//!
+//! The default maximum number of concurrent worker threads is controlled by the
+//! `analysis.limits.workerThreadCount` setting but can be adjusted at runtime via
+//! [set_worker_thread_count].
+//!
+//! The worker threads are native threads, managed by the Binary Ninja core. If
+//! more control over the thread is required, consider using the
+//! [crate::backgroundtask::BackgroundTask] class.
+
+use core::{ffi, mem, ptr};
+
+use binaryninjacore_sys::*;
+
+use crate::string::BnStrCompatible;
+
+#[repr(transparent)]
+pub struct MainThreadAction {
+    handle: ptr::NonNull<BNMainThreadAction>,
+}
+
+impl Clone for MainThreadAction {
+    fn clone(&self) -> Self {
+        let result = unsafe { BNNewMainThreadActionReference(self.as_raw()) };
+        unsafe { Self::from_raw(ptr::NonNull::new(result).unwrap()) }
+    }
+}
+
+impl Drop for MainThreadAction {
+    fn drop(&mut self) {
+        unsafe { BNFreeMainThreadAction(self.as_raw()) }
+    }
+}
+
+impl MainThreadAction {
+    pub(crate) unsafe fn from_raw(handle: ptr::NonNull<BNMainThreadAction>) -> Self {
+        Self { handle }
+    }
+
+    pub(crate) unsafe fn ref_from_raw(handle: &*mut BNMainThreadAction) -> &Self {
+        assert!(!handle.is_null());
+        mem::transmute(handle)
+    }
```

Comment:
Is this function actually used anywhere? And since `*mut T: Copy`, you could reasonably just call `from_raw` instead and get an owned `Self`. Plus, elsewhere in the codebase, the return type of `ref_from_raw` is usually `Ref<Self>`. I also don't like the use of `transmute` here.

Reply:
This function is used by the `cb_add_action`. The copy is intentionally avoided here, to avoid problems like #5496

Reply:
In that case, a better alternative to raw transmute is something like:
```rust
unsafe { &*(handle as *const _ as *const Self) }
```
It's more clear what's going on here and it doesn't tap into transmute's full power (which is a good thing)